src/AccessLogEntry.h

```diff
@@ -177,7 +179,8 @@ class AccessLogEntry: public RefCountable
     SBuf lastAclData; ///< string for external_acl_type %DATA format code
 
     HierarchyLogEntry hier;
-    HttpReply *reply = nullptr;
+    HttpReplyPointer reply; ///< sent reply (SeeSquidWiki)
```

Comment:
All this fussing over the terminology seems a bit unnecessary to me. Some of it appears to be derived from the fact that this _single_ member is not really suitable for top-level ALE placement relative to the logformat needs and the 'h' and 'ha' codes specifically have not yet been updated to use it.

We really have three types of response being juggled;
 1) the "reply from server or peer" which really should be at ALE::HttpDetails::serverReply
 2) the "adapted reply" - partially processed, maybe adapted from (1) but not yet sent as (3).
   - this one maybe could be top-level, but in a member named better than just "reply". or for now stored as (3).
 3) the "reply sent to the client" which really should be at ALE::HttpDetails::clientReply

The "<h" is (1), "<ha" is (3).

Reply:
> The "<h" is (1), "<ha" is (2) / (3).

To clarify: we don't have '<ha'.
'<h' logs sent replies (as it logged before the PR changes).

Reply:
That was part of my point. Much of this mind-bending to re-define the common terms you guys seem to have been doing is working around that. After it gets added there is likely very little remainder to worry about.

Reply:
> The "<h" is (1), "<ha" is (2) / (3).

Ideally, the reply terminology and log format codes should correlate with their request counterparts. However, '>ha' stands for 'received request after adaptation/redirection' and usually differs from the request, whereas you suggest that '<ha' should keep the sent reply( if available). So, in order to simplify/unify this task, we would need to have three logformat codes for request (virgin/adapted/sent) and three for reply. Currently we have only two for request(virgin/adapted) and one for reply (sent). Do you think we should add the remaining(missing) logformat codes and (if yes) do it in this PR?

Reply:
> Amos: After it gets added there is likely very little remainder to worry about.

Regardless of how many %codes we add or have, we still need to explain what those codes mean. At least some of those %codes (and fields/methods) will require lengthy, non-trivial explanations because their computation method is _naturally_ complex. For example, many admins want to log the "latest reply available", which is a complex notion regardless of how many (other) %codes we have/add.

So yes, there will be quite a bit to "worry about" even after adding more %codes. And, ideally, deciding what %codes to add and thinking about how to describe them should be done at the same time, not in sequence!

Your ad-hoc attempt to suggest what should be done instead is a good illustration of the natural problem complexity -- the suggestion immediately spawned a discussion full of controversial ideas and misunderstandings (all using inadequate terminology). If we want to revise message %codes, a _comprehensive_ solution should be carefully thought out outside a largely unrelated PR comment thread.

> Eduard: Do you think we should add the remaining(missing) logformat codes and (if yes) do it in this PR?

The answer to this question is simple: We should remove documentation changes from this PR if their presence triggers "add more %codes" change requests. In this case, improving documentation simply becomes too much side work. We should cut our losses and just focus on fixing the bug, leaving the existing documentation as useless as it is right now.

@yadij, if you do not like the idea of adding SeeSquidWiki references, we will remove those controversial documentation improvements from this PR. Is that your preference?

Documentation and %codes issues can be revisited in the future, of course.

----

The only viable alternative to so called "redefining of common terms" is to invent uncommon ones.
* If we continue to restrict this PR from adding new %codes, then I do not see any good uncommon terms that we should use instead. Referring to wiki where the current status quo can be documented and where the terminology can be developed/perfected is an imperfect but viable solution here IMO. However, we do not have to change documentation at all if there is no consensus regarding this SeeSquidWiki idea.
* If we add new %codes, then, for _some_ sets of new %codes, those %codes _themselves_ can be used as "uncommon terms". However, doing this right is too much optional work. We will not do it in this PR.

Reply:
> Amos: After it gets added there is likely very little remainder to worry about.
> 
> Regardless of how many %codes we add or have, we still need to explain what those codes mean. At least some of those %codes (and fields/methods) will require lengthy, non-trivial explanations because their computation method is naturally complex. For example, many admins want to log the "latest reply available", which is a complex notion regardless of how many (other) %codes we have/add.

Years of experience with the request-path format codes belies those assertions. Request handling has just as much processing applied and the descriptions are still mostly one-liners which are easily read and understood without having to resort to diagrams and pages teaching admin definitions. The exceptions have been where you chose to add new un-common terms.

The major relevant difference between the two right now is that the response code paths are lacking a lot of the cleanup which happened in the request pathways already. Which is what my initial comment was saying, there are things here which are already planned out to change. When it happens that change can be accompanied by a better description. It is a waste of time re-doing the description now when things are still a mess ...

> So yes, there will be quite a bit to "worry about" even after adding more %codes. And, ideally, deciding what %codes to add and thinking about how to describe them should be done at the same time, not in sequence!

By that reason ideally the existing %code should not have the apparently arbitrary description change we see going on here.

The %codes I mentioned in my comment (and other missing %codes too) have already been selected when the existing terms were designed. The label reservations can be found in ByteCode.h as commented out entries symmetrical with request handling codes (though the Token.cc "<ha" part seems to be missing now):
``

    /* response header details pre-adaptation */
    LFT_REPLY_HEADER,
    LFT_REPLY_HEADER_ELEM,
    LFT_REPLY_ALL_HEADERS,

    /* response header details post-adaptation */
    /* LFT_ADAPTED_REPLY_HEADER, */
    /* LFT_ADAPTED_REPLY_HEADER_ELEM, */
    /* LFT_ADAPTED_REPLY_ALL_HEADERS, */
``

All that remains is the code cleanup in the reply handling to provide the pointers to ALE differently, and implementation of the %code's using the resulting ALE members. Sadly the incremental changes leading to that end kept getting stalled in review over not being important enough in small chunks.


> Your ad-hoc attempt to suggest what should be done instead 

Is not ad-hoc at all. It is the result of a) the previous round of discussions the then-core team had on this topic, and b) years of study and months of drawn out discussions about the design for ALE and MasterTransaction contents. The docs terms are well established and have been used by all the logformat additions up until this point. Now I see the two of you without any other consultation beginning a complete re-write to naming things in a way that looks quite at odds with the previous long-term plans.

> If we want to revise message %codes, a comprehensive solution should be carefully thought out outside a largely unrelated PR.

Agreed.

> Eduard: Do you think we should add the remaining(missing) logformat codes and (if yes) do it in this PR?
> 
> The answer to this question is simple: We should remove documentation changes from this PR if their presence triggers "add more %codes" change requests. In this case, improving documentation simply becomes too much side work. We should cut our losses and just focus on fixing the bug, leaving the existing documentation as useless as it is right now.

> @yadij, if you do not like the idea of adding SeeSquidWiki references, we will remove those controversial documentation improvements from this PR. Is that your preference?

It is, yes.

Reply:
> we will remove those controversial documentation improvements

Done at 7bd1b13.

Reply:
> Years of experience with the request-path format codes belies those assertions.

That is your opinion, not a fact. My years of experience have been different. Glad we may have found a way to move forward despite these differences in opinion. Are you OK with the current documentation changes in this PR? If not, we should remove them (which is what I meant earlier but @eduard-bagdasaryan has removed just the SeeSquidWiki references).